nifi-nar-bundles/nifi-oauth-bundle/nifi-oauth/src/main/java/org/apache/nifi/oauth/AbstractOAuthControllerService.java

```diff
@@ -0,0 +1,172 @@
+
+    public static final PropertyDescriptor EXPIRE_TIME_SAFETY_NET = new PropertyDescriptor
+            .Builder().name("Expire time safety net in seconds")
+            .displayName("Expire time safety net in seconds")
+            .description("There can be a chance that the authentication server and the NiFi agent clocks could be out of sync. Expires In is an OAuth standard " +
+                    "that tells when the access token received will be invalidated. Comparing the current system clock to that value and understanding if the " +
+                    "access token is still valid can be achieved this way. However since the clocks can be out of sync with the authentication server " +
+                    "this property ensures that the access token can be refreshed at a configurable interval before it actual expires to ensure no downtime " +
+                    "waiting on new tokens from the authentication server.")
+            .defaultValue("10")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    protected void onEnabled(final ConfigurationContext context)
+            throws InitializationException {
+        Map<PropertyDescriptor, String> allProperties = context.getProperties();
+        Iterator<PropertyDescriptor> itr = allProperties.keySet().iterator();
+        while (itr.hasNext()) {
+            PropertyDescriptor pd = itr.next();
+            if (pd.isDynamic()) {
+                extraHeaders.put(pd.getName(), allProperties.get(pd));
+            }
+        }
+
+        this.authUrl = context.getProperty(AUTH_SERVER_URL).getValue();
+        this.expireTimeSafetyNetSeconds = new Long(context.getProperty(EXPIRE_TIME_SAFETY_NET).getValue());
+
+        // Name of the OAuth2 spec fields that should be extracted from the JSON authentication response. While in theory every provider should be
+        // using the same names in the response JSON that rarely happens. These values are used to ensure that if the provider does NOT follow the
+        // spec the user can still use this integration approach instead of having to write something custom.
+        this.accessTokenRespName = context.getProperty(RESPONSE_ACCESS_TOKEN_FIELD_NAME).getValue();
+        this.expireTimeRespName = context.getProperty(RESPONSE_EXPIRE_TIME_FIELD_NAME).getValue();
+        this.expireInRespName = context.getProperty(RESPONSE_EXPIRE_IN_FIELD_NAME).getValue();
+        this.tokenTypeRespName = context.getProperty(RESPONSE_TOKEN_TYPE_FIELD_NAME).getValue();
+        this.scopeRespName = context.getProperty(RESPONSE_SCOPE_FIELD_NAME).getValue();
+    }
+
+    public boolean isOAuthTokenExpired() {
+        if (expiresTime == -1 && expiresIn == -1)
+            return true;
+
+        if (expiresTime > 0) {
+            // Use the actual time that the token will authenticate
+            if ((expiresTime - (expireTimeSafetyNetSeconds * 1000)) <= System.currentTimeMillis()) {
```

Comment:
Let's say `System.currentTimeMillis()` returns a static value *1000* representing "now" and `expiresTime` is set by the remote system and returns a static value *3000* (2 seconds after *now*). With the default value of `expireTimeSafetyNetSeconds` as *10*, if the clock jitter is non-existent or "positive" (i.e. the local system's clock is "ahead"/returns a higher value than the remote system), this will evaluate to `3000 - (10 * 1000) <= 1000` which is `-7000 <= 1000` and eventually `true`. However, if the clock jitter is "behind", for example "now" on the remote system is *9001* while it's only *1000* locally (8.001 seconds difference, which should be within the "safety net" range), the expression will be `11001 - (10 * 1000) <= 1000` which is `1001 <= 1000` which is false. I believe this needs to use absolute value evaluations to handle both positive and negative offsets.

Reply:
Andy. I think I could have chosen a better name for the "expireTimeSafetyNetSeconds" variable. My intended behavior was for that variable to be more of a method for causing the auth token to be refresh before the token was actually expired. My thought process around this was if users could control having their token refreshed before they were actually invalid it would allow them a rough way to help prevent bottlenecking of processing when the token was invalid and the processors using the controller service were having to wait on the token to be refreshed. Does that make sense?

Reply:
Jeremy, sorry, I don't think I'm following. I can understand the desire to prevent token expiration, but I don't understand how that maps directly here or how the current code handles the scenario I described above.

Reply:
Jeremy and I discussed and since the fail state is that the CS just tries to reauthenticate, this variable and logic will be removed completely.